Add How does ML.NET work?

natke opened the pull request:
## Summary

Add new article: https://review.docs.microsoft.com/en-us/dotnet/machine-learning/how-does-mldotnet-work?branch=pr-en-us-11871

[AB#1479484](https://dev.azure.com/mseng/677da0fb-b067-4f77-b89b-f32c12bb8617/_workitems/edit/1479484)

Note: art work is preliminary but please give feedback on its content.

ETA: there are currently two broken links in this PR to the load data and train model how-tos. I'm assuming that @luisquintanilla's PRs will get merged before this one.

natke committed:
Add How does ML.NET work - first draft

natke committed:
Second draft, with some draft diagrams.

natke committed:
Add link for single predictions

natke committed:
Add redirect for changed filename

natke committed:
Add new diagram; fix code indentation; run acrolinx

natke committed:
Update image + format

natke committed:
Fix links

natke committed:
Fix Preview link

natke committed:
Tidy code blocks and wording

natke committed:
Clarify object types

natke committed:
Update regression image

natke committed:
Fix formula

natke committed:
Fix indent

natke committed:
Fix code

natke committed:
Update title

natke committed:
Remove index.md

natke committed:
Fixed redirection casing

natke committed:
Fix internal anchors to old doc

natke committed:
Fix more anchors

natke committed:
Add data entry to the glossary

natke committed:
Anchor fix

JRAlexander commented on docs/machine-learning/how-does-mldotnet-work.md:
```diff
@@ -0,0 +1,262 @@
+
+### More complex
+
+A more complex model classifies financial transactions into categories using the text description of the transaction.
```
A more complex model classifies financial transactions into categories using the transaction text description.

JRAlexander commented on docs/machine-learning/how-does-mldotnet-work.md:
```diff
@@ -0,0 +1,262 @@
+
+A more complex model classifies financial transactions into categories using the text description of the transaction.
+
+Each transaction description is broken down into a set of features by removing redundant words and characters, counting word and character combinations, then training a linear model based on the set of categories in the training data. The more similar a new description is to the ones in the training set, the more likely it will be assigned to the same category. 
```
Maybe something like: "Each transaction description is broken down into a set of features by removing redundant words and characters, and counting word and character combinations. The feature set trains a linear model based on the set of categories in the training data."

JRAlexander commented on docs/machine-learning/how-does-mldotnet-work.md:
````diff
@@ -0,0 +1,262 @@
+        // RMS error: 0.09
+```
+
+The evaluation metrics tell us that the error is low-ish, and that correlation between the predicted output and the test output is high. That was easy! In real examples, it takes more tuning to achieve good model metrics.
````
"us" -> "you". Use second person.

JRAlexander commented on docs/machine-learning/how-does-mldotnet-work.md:
```diff
@@ -0,0 +1,262 @@
+
+## Architecture
+
+In this section, we go through the architectural patterns of ML.NET. If you are an experienced .NET developer, some of these patterns will be familiar to you, and some will be less familiar. Hold tight, while we dive in!
```
"we" -> "you".

JRAlexander commented on docs/machine-learning/how-does-mldotnet-work.md:
```diff
@@ -0,0 +1,262 @@
+
+### Fit the model
+
+Once the objects in the pipeline have been created, data can be used to train or fit the model.
```
Would this be better? "Once the objects in the pipeline have been created, data can be used to fit (train) the model."

natke replied:
Removed fit altogether

JRAlexander commented on docs/machine-learning/how-does-mldotnet-work.md:
````diff
@@ -0,0 +1,262 @@
+    var model = pipeline.Fit(trainingData);
+```
+
+Calling `Fit()` uses the input training data to estimate the parameters of the model. Remember in our linear regression model above, we had two model parameters: **bias** and **weight**. After the `Fit()` call, the values of the parameters are known. Most models will have many more parameters than this.
````
Maybe something like this: "Calling `Fit()` uses the input training data to estimate the parameters of the model. This is known as training the model."

natke replied:
Good suggestion

JRAlexander commented on docs/machine-learning/how-does-mldotnet-work.md:
````diff
@@ -0,0 +1,262 @@
+    var pipeline = mlContext.Transforms.Concatenate("Features", new[] { "Size" })
+ ```
+
+All algorithms also create new columns after they have performed a prediction. The fixed names of these new columns depend on the type of machine learning algorithm. For the regression task, one of the new columns is called **Score**. This is why we attributed our price data with this name.
````
"This is why you attributed your price data with this name."

JRAlexander commented on docs/machine-learning/how-does-mldotnet-work.md:
````diff
@@ -0,0 +1,262 @@
+    var debug = testPriceDataView.Preview();
+```
+
+You can watch the `debug` variable in the debugger and examine its contents. Do not use the Preview method in production code, as it significantly degrades performance.
````
"Do not" -> "Don't"

JRAlexander commented on docs/machine-learning/how-does-mldotnet-work.md:
```diff
@@ -0,0 +1,262 @@
+
+![Text Classification Model](./media/text-classification-model.png)
+
+Both the house price model and the text classification model are **linear** models. Depending on the nature of your data and the problem you are solving, you can also use **decision tree** models, **generalized additive** models and others. You can find out more about the models in [How to choose an ML.NET model](./resources/tasks.md).
```
"you are" -> "you're" and there should be a comma after "**generalized additive** models".

JRAlexander commented on docs/machine-learning/how-does-mldotnet-work.md:
```diff
@@ -0,0 +1,262 @@
+
+## Data preparation
+
+In most cases, the data that you have available in your application is not suitable to be used directly to train a machine learning model. The raw data needs to be prepared, or pre-processed before it can be used to find the parameters of your model. Your data may need to be converted from string values to a numerical representation. You might have redundant information in your input data. You ma ...
```
"the data that you have available in your application " -> "your application data"

JRAlexander replied:
"is not" -> "isn't"

natke replied:
removed in your application. Application data sounds more like environment variables and the like

JRAlexander commented on docs/machine-learning/how-does-mldotnet-work.md:
```diff
@@ -0,0 +1,262 @@
+
+Each transformation in the pipeline has an input schema (data names, types, and sizes that the transform expects to see on its input); and an output schema (data names, types, and sizes that the transform produces after the transformation). 
+
+If the output schema from one transform in the pipeline does not match the input schema of the next transform, ML.NET will throw an exception.
```
"does not" > "doesn't"

JRAlexander commented on docs/machine-learning/how-does-mldotnet-work.md:
```diff
@@ -0,0 +1,262 @@
+
+If the output schema from one transform in the pipeline does not match the input schema of the next transform, ML.NET will throw an exception.
+
+A data view object has columns and rows. Each column has a name and a type and a length. For example: the input columns in the house price example are **Size** and **Price**. They are both type <xref:System.Single> and they are scalar quantities rather than vector ones.
```
"Each column name has a name, type, and length"

natke replied:
?

JRAlexander commented on docs/machine-learning/how-does-mldotnet-work.md:
```diff
@@ -0,0 +1,262 @@
+
+   ![ML.NET Data View example with house price prediction data](./media/ml-net-dataview.png)
+
+All ML.NET algorithms look for an input column that is a vector. By default this vector column is called **Features**. This is why we concatenated the **Size** column into a new column called **Features** in our house price example.
```
"This is why the **Size** column was concatenated into a new column called **Features** in the house price example."

JRAlexander reviewed (changes requested):
Looks good. A few small items.

natke committed:
Update artwork

natke commented on docs/machine-learning/how-does-mldotnet-work.md:
```diff
@@ -0,0 +1,262 @@
+
+- data loading and saving: <xref:Microsoft.ML.Data>
+- data preparation :<xref:Microsoft.ML.TransformsCatalog>
+- algorithms for model training
```
Where would you link to here @luisquintanilla?

luisquintanilla replied:
Hey, saw you made the change. The links you used were the ones I was referring to.

natke commented on docs/machine-learning/how-does-mldotnet-work.md:
```diff
@@ -0,0 +1,262 @@
+
+You can learn more about model training in [How to train your model](./how-to-guides/train-model.md)
+
+The resulting model object implements the <xref:Microsoft.ML.ITransformer> interface. That is, the model transforms transforms input data into predictions. 
```
Good catch, that's a typo

natke commented on docs/machine-learning/how-does-mldotnet-work.md:
```diff
@@ -0,0 +1,262 @@
+
+You can find out more about output columns of different machine learning tasks in the [Machine Learning Tasks](resources/tasks.md) guide.
+
+An important property of DataView objects is that they are evaluated **lazily**. Data views are only loaded and operated on during model training and evaluation, and data prediction. While you are writing and testing your ML.NET application, you can use the Visual Studio debugger to take a peek at any data view object by calling the [Preview](xref:Microsoft.ML.DebuggerExtensions.Preview*) method.
```
I was actually trying to group model training and evaluation, and keep prediction separate. Any other suggestions to keep this intent?

natke commented on docs/machine-learning/how-does-mldotnet-work.md:
```diff
@@ -0,0 +1,262 @@
+
+Examples of the type of prediction that you can make with ML.NET include:
+
+- Categorization, or classification. For example: automatically divide customer feedback into positive and negative categories
```
Great suggestion!

natke commented on docs/machine-learning/how-does-mldotnet-work.md:
```diff
@@ -0,0 +1,262 @@
+
+![Linear Regression Model with bias and weight parameters](./media/linear-regression-model.svg)
+
+The model is simply: `Price = b + Size * w`. The parameters `b` and `w` are estimated by fitting a line on a set of (size, price) pairs. The data used to find the parameters of the model is called **training data**. The inputs of a machine learning model are called **features**. In this example, `Size` is the only feature. The ground-truth values used to train a machine learning model are called  ...
```
In this case, this is what the model is. Do you think it is confusing?

natke committed:
Update after review

natke committed:
Refinements after review

natke committed:
Link directly to catalogs

natke committed:
Refine title

natke committed:
Refine definitions

natke committed:
Add latex formatting

natke committed:
Add correct filenames

natke committed:
Fix typos

natke commented:
@JRAlexander and @luisquintanilla, I've updated after review. I addressed *most* of the comments with your suggestions. Some I didn't. We can discuss if you like. Can you please re-review?

There is one conceptual diagram to come: the code workflow one.

luisquintanilla commented on docs/machine-learning/how-does-mldotnet-work.md:
````diff
@@ -0,0 +1,276 @@
+
+In real-life applications, your model training and evaluation code will be separate from your prediction. In fact, these two activities are often performed by separate teams. Your model development team can save the model for use in the prediction application.
+
+```csharp   
````
Suggestion: This snippet can be simplified since the model is being saved to a local file. Use the method that takes in the string path as opposed to a Stream.

```csharp
mlContext.Model.Save(model, trainingData.Schema,"model.zip");
```

natke replied:
Well-spotted. Fixing, thanks!

natke committed:
More updates after review + refactor intro article

natke committed:
Fix TOC

natke committed:
Simplify save model code snippet
